sdk/core/azure-core/inc/azure/core/http/http.hpp

```diff
@@ -91,22 +91,76 @@ namespace Azure { namespace Core { namespace Http {
   };
 
   /**
-   * HTTP request method.
-   *
+   * @brief The method to be performed on the resource identified by the Request.
    */
   class HttpMethod final {
   public:
     HttpMethod() = delete;
+
+    /**
+     * @brief Constructs a new HttpMethod from a given string.
+     *
+     * @note Won't check if \p value is a known HttpMethod defined as per any RFC.
+     *
+     * @param value A given string to represent the HttpMethod.
+     */
     explicit HttpMethod(std::string value) : m_value(std::move(value)) {}
+
+    /**
+     * @brief Compares two instances of `HttpMethod` for equality.
+     *
+     * @param other Some `HttpMethod` instance to compare with.
+     * @return `true` if instances are equal; otherwise, `false`.
+     */
     bool operator==(const HttpMethod& other) const { return m_value == other.m_value; }
+
+    /**
+     * @brief Compares two instances of `HttpMethod` for equality.
+     *
+     * @param other Some `HttpMethod` instance to compare with.
+     * @return `false` if instances are equal; otherwise, `true`.
+     */
     bool operator!=(const HttpMethod& other) const { return !(*this == other); }
+
+    /**
+     * @brief Returns the HttpMethod represented as a string.
+     */
```

Comment:
Does this create a new string or a copy, every time? Should the caller store it in a local, or is it OK to keep calling `ToString()` in a loop?

Reply:
This is defined by the method signature and not by docs:

const std::string& ToString() ->  Returns a const ref to some state (not creating a new string). Caller can read the state but not mutate it

std::string& ToString() -> Caller can access the state and mutate it.

std::string ToString() -> Returns a new string on every call.


If the caller want to make a copy of the value read from the instance's state, they would do `auto copyName = instance.ToString();`  After the const ref is returned, the compiler makes a copy into `copyName`.  Otherwise, the caller would do `auto& readOnly = instance.ToString()` to access the memory and value without making copies

Reply:
I would not document it, and it is better for the customer code, if the performance here is important, to cache it once before the loop without making assumption how effective the SDK implementation is. Today we may return const reference, tomorrow we restructure internals and start returning a string by value. But if we made promises in documentation, we won't be able to do that.

Reply:
Should we add some template remarks to all such const ref returning methods (or maybe specifically the string returning ones), to let novice users know that `auto copyName = instance.ToString();` will create a hidden copy, and you should generally start off as: `auto& readOnly = instance.ToString()`, unless there is a reason to create a copy?

Reply:
> I would not document it, and I'd say, if the performance here is important to the customer, cache the result of the function once before the loop without making assumption how effective the SDK implementation is. Today we may return const reference, tomorrow we restructure internals and start returning a string by value. But if we made promises in documentation, we won't be able to do that.

Wouldn't that require a breaking change/change of signature, which we wouldn't do anyway?

Reply:
I would not document the language itself, and I don't think that suggesting to use `auto&` is a good advice, unless it is in a perf bottleneck. Reference to something you don't control the lifetime has greater potential for bugs. I'd say, if you are a novice user, do a copy, it works better most of the time. Compiler will do good optimizations. If you are an expert, also use copy, unless you know you are in perf bottleneck. I suspect that `auto` type works this way in C++ for exactly this reason: it was added to the language later, in C++11, and the goal probably was that `string x = f()` and `auto x = f()` has the same meaning and type, because it is common pattern with less problems and assumptions. If an exact type is needed, other option is to use `decltype(f())` in place of auto, then your code is guaranteed to use reference when function returns reference, and may create a copy when it returns value. But I would leave that to compiler to do optimization in most cases.